crates/chain-state/src/in_memory.rs

```diff
@@ -723,6 +721,55 @@ impl BlockState {
     pub fn iter(self: Arc<Self>) -> impl Iterator<Item = Arc<Self>> {
         std::iter::successors(Some(self), |state| state.parent.clone())
     }
+
+    /// Tries to find a block by [`BlockHashOrNumber`] in the chain ending at this block.
+    pub fn block_on_chain(&self, hash_or_num: BlockHashOrNumber) -> Option<&Self> {
+        self.chain().find(|block| match hash_or_num {
+            BlockHashOrNumber::Hash(hash) => block.hash() == hash,
+            BlockHashOrNumber::Number(number) => block.number() == number,
+        })
+    }
+
+    /// Tries to find a transaction by [`TxHash`] in the chain ending at this block.
+    pub fn transaction_on_chain(&self, hash: TxHash) -> Option<TransactionSigned> {
+        for block_state in self.chain() {
+            if let Some(tx) =
+                block_state.block_ref().block().body.transactions().find(|tx| tx.hash() == hash)
+            {
+                return Some(tx.clone())
+            }
+        }
+        None
+    }
```

Comment:
Maybe I'm wrong here, but something like this could be more compact, wdyt?

```rust
/// Tries to find a transaction by [`TxHash`] in the chain ending at this block.
pub fn transaction_on_chain(&self, hash: TxHash) -> Option<TransactionSigned> {
    self.chain().find_map(|block_state| block_state.block_ref().block().body.transactions().find(|tx| tx.hash() == hash).cloned())
}
```